### PR TITLE
Feature/crypto

### PR DESCRIPTION
**Pull request by ganfra**
This PR brings crypto to RiotX.
It allows to decrypt/encrypt messages, use SAS verification and keys backup (only from settings screen).
It has some limitations: 
- Failures are not yet well handled
- It does send encrypted messages even when unknown devices are in the room (without warning).
- There are probably some other issues
- There are still code to clean up

### CODE REVIEW

File: matrix-sdk-android/build.gradle [hunk 1]
```diff
@@ -45,8 +45,8 @@ android {
 
         debug {
             // Set to true to log privacy or sensible data, such as token
+            // TODO Set to false
```
Comment (BillCarsonFr) [inline matrix-sdk-android/build.gradle]: done?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/events/model/Event.kt [hunk 2]
```diff
@@ -81,10 +84,147 @@ data class Event(
...
+     * @return true if this event is encrypted.
+     */
+    fun isEncrypted(): Boolean {
+        return TextUtils.equals(type, EventType.ENCRYPTED)
+    }
+
+    /**
+     * Update the clear data on this event.
+     * This is used after decrypting an event; it should not be used by applications.
+     * It fires kMXEventDidDecryptNotification.
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/events/model/Event.kt]: Does it?

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/actions/EnsureOlmSessionsForDevicesAction.kt [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+
+    suspend fun handle(devicesByUser: Map<String, List<MXDeviceInfo>>): Try<MXUsersDevicesMap<MXOlmSessionResult>> {
+        val devicesWithoutSession = ArrayList<MXDeviceInfo>()
+
+        val results = MXUsersDevicesMap<MXOlmSessionResult>()
+
+        val userIds = devicesByUser.keys
+
+        for (userId in userIds) {
```
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/actions/EnsureOlmSessionsForDevicesAction.kt]: I see a lot of potential NPE code. Maybe we could reduce usage of !!

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/events/model/Event.kt (hunk 2 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/events/model/Event.kt]: aha, iOS comment copy/pasted twice!

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/actions/EnsureOlmSessionsForDevicesAction.kt (hunk 3 above)
Comment (ganfra) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/actions/EnsureOlmSessionsForDevicesAction.kt]: Yes, it's converting java to kotlin which leads to this, but it should be clean.

**Pull request merged**